### ISSUE TITLE
[QST] cudf.read_csv 4GB.txt on a 8GB GTX1070 returns cudaErrorMemoryAllocation out of memory

### ISSUE THREAD

**Issue by edgecate** [rapidsai/cudf#8653]
TLDR - why can I only load a 2GB CSV on an 8GB GTX1070?

I'm using an 8GB GTX1070 for CUDF.

I've created a bunch of CSV files (1GB, 2GB, 4GB, 8GB) with a .TXT extension but they are comma separated.

My napkin math assumption was that since I have 8GB of GPU memory, I would be able to load my 4GB or 8GB files into a CUDF.

However, the maximum I can load is 2GB. 

If I try the 4GB or 8GB file, I get an Out of Memory error:
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-9-4945b801b398> in <module>
      1 # Read CSV
----> 2 df = cudf.read_csv('04GB Data.txt')

/opt/conda/envs/rapids/lib/python3.8/contextlib.py in inner(*args, **kwds)
     73         def inner(*args, **kwds):
     74             with self._recreate_cm():
---> 75                 return func(*args, **kwds)
     76         return inner
     77 

/opt/conda/envs/rapids/lib/python3.8/site-packages/cudf/io/csv.py in read_csv(filepath_or_buffer, lineterminator, quotechar, quoting, doublequote, header, mangle_dupe_cols, usecols, sep, delimiter, delim_whitespace, skipinitialspace, names, dtype, skipfooter, skiprows, dayfirst, compression, thousands, decimal, true_values, false_values, nrows, byte_range, skip_blank_lines, parse_dates, comment, na_values, keep_default_na, na_filter, prefix, index_col, **kwargs)
     68         na_values = [na_values]
     69 
---> 70     return libcudf.csv.read_csv(
     71         filepath_or_buffer,
     72         lineterminator=lineterminator,

cudf/_lib/csv.pyx in cudf._lib.csv.read_csv()

MemoryError: std::bad_alloc: CUDA error at: /opt/conda/envs/rapids/include/rmm/mr/device/cuda_memory_resource.hpp:69: cudaErrorMemoryAllocation out of memory

nvidia-smi shows 5221MB of GPU CUDF memory usage with other processes only taking up 336MB, so I should have at least 2GB spare.

(base) andrew@Ubuntu:~$ nvidia-smi
Sun Jul  4 12:03:47 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.42.01    Driver Version: 470.42.01    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:07:00.0  On |                  N/A |
|  0%   55C    P0    57W / 200W |   5710MiB /  8111MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1162      G   /usr/lib/xorg/Xorg                188MiB |
|    0   N/A  N/A      1343      G   /usr/bin/gnome-shell              104MiB |
|    0   N/A  N/A      3134      G   ...AAAAAAAAA= --shared-files       44MiB |
|    0   N/A  N/A      3547      C   ...da/envs/rapids/bin/python     5221MiB |
+-----------------------------------------------------------------------------+

[nvidia-smi.txt](https://github.com/rapidsai/cudf/files/6759278/nvidia-smi.txt)

**Comment by beckernick** [comment]
Hi @edgecate . In general, reading a file requires allocating more memory than the size of the file itself. It's possible that in this case your GPU does not have enough memory for data of that size.

For out-of-memory workloads that require spilling data between the GPU and CPU, we recommend using Dask with cuDF. Hopefully, [this blog](https://medium.com/rapids-ai/reading-larger-than-memory-csvs-with-rapids-and-dask-e6e27dfa6c0f) can provide some guidance

**Comment by edgecate** [comment]
Thank-you, Nick! :)
